samtranslator/swagger/swagger.py

```diff
@@ -781,6 +782,42 @@ def _set_method_apikey_handling(self, path, method_name, apikey_required):
             if security != existing_security:
                 method_definition["security"] = security
 
+    def add_request_validator_to_method(self, path, method_name, validate_body=False, validate_request=False):
+        """
+        Adds request model body parameter for this path/method.
+
+        :param string path: Path name
+        :param string method_name: Method name
+        :param bool validate_body: Add validator parameter on the body
+        :param bool validate_request: Validate request
+        """
+
+        normalized_method_name = self._normalize_method_name(method_name)
+        normalized_path_name = SwaggerEditor.get_path_name_normalized(path)
+        validator_name = "{normalized_path_name}-{normalized_method_name}-validator".format(
```

Comment:
So looking at this now, I wonder if this is the right design for the spec. 

Thought process:
This will add N number of validators to the `x-amazon-apigateway-request-validators` part of the doc. But in reality there are only 3 combinations to `x-amazon-apigateway-request-validators`: both request body and request param, only request body, only request param. Sure it is possible for API Gateway to add other validators here but my assumption is that it is still finite and small. This leads to larger template files and OpenApi definitions when it isn't needed. I wonder if we should have a way to define a set of "validator combinations" instead. 

Proposal/discussion:
I think all the main logic is mostly the same, but if we agree deduping is better maybe what we really need to do here is:
* Add a dict of Validators to the `AWS::Serverless::Api` (and maybe Http?) that accepts a dict of key to validators. Bascially this dict of key values would be added directly to `x-amazon-apigateway-request-validators` in the OpenApi Definition.
* This new property would be added to Globals to support not having to write a `AWS::Serverless::Api` resources at all and let SAM generate this from the API Events in `AWS::Serverless::Function` (implicit API).
* Add a `ValidatorName` type property that will be used to specify in the API Event to key into the new property added above.

I think this might better align with what API Gateway already defines in OpenAPI Definitions. Happy to have an open dialog about this and sorry this is happening now vs catching it during the proposal. I didn't realize API Gateway defined it in this manner, which is my bad.

Reply:
Hi @jfuss I actually thought in the way let it explicitly to the path/method has a validator, but haven't thought about the small limit of combinations (sorry the correction, but would be 4 combinations). I think, maybe defining a pre-defined combination on a constant definition, and add only when called, ie:

FULL = (request, body)
BODY = (body, !request)
REQUEST = (!body, request)
NO_VALIDATE =(!request, !body)

If not defined
NONE = ()

Each time we have a iteration, I would try to get which one is being requested then add it to the document. A second time, if the definition already exist I just re-use it don't add a new one. If is not defined, I do not add it to the spec.

Tbh, I think leaving those 2 options simple as a boolean would be easier to understand and to use. Let me know what you think about it!

Reply:
So, about this issue, I have changed my PR to that limit of combinations. Where if the spec has been defined already, I don't add a new one with a diff name instead just add what is already defined. I think that makes more sense, would simplify implementations and let easier to the client point of view as adding it to the validatorName is already possible doing Body Definitions, which is a bit more complex than only add those 2 parameters.

I added a testcase where If the combination is repeated I am not adding a new one, but re-using an existent one already defined on the swagger doc.

Let me know your thinking around that.

Reply:
@rohandalvi You are correct, there are 4 combinations. I ignored the `NO_VALIDATE` case and assumed having `False` set to both Validates was equivalent. I can see value in having this still to allow customers the ability to audit the template and ensure all paths have this set.

What you have/proposed does work, though my main concern is around keeping this updated in the future. SAM becomes the bottle neck in this solution. While this might not ever happen, if API Gateway adds another "validate" property, SAM needs to find a way to handle this. We could keep the updating/adding these pre-defined configuration but SAM will always be behind and we risk creating a more confusing Spec. For example `FULL` would need to be expanded have any of the "new validation" which would be a breaking change for SAM (an updated will change the behavior of customer's template without any new being added in the Spec). Now, we could not do that but then it's hard to tell what `FULL` is really for. This will also exponentially increase the number of combinations to support, which is fine in small numbers but could become large as time goes on.

Given that, my opinion right now is to enable this is a way that SAM can be out of the way in the future. I like the simplicity of your solution but the possible maintenance is what I fear. Let me ponder this a little more and bring it up with my team to see what we collectively think. I think this comes down to a balancing act between simplicity and longevity/maintenance.